### ISSUE TITLE
PARI's qfbclassno() may give incorrect results

### ISSUE THREAD

**Issue by yyyyx4** [sagemath/sage#34801]
<div id="comment:0"></div>

Since #23986, Sage calls PARI's `qfbclassno()` to compute the class number of a quadratic order. However, the documentation for that PARI function says:

> [...] this function only implements part of Shanks's method (which allows to speed it up considerably). It gives unconditionnally correct results for |D| < 2.10¹⁰, but may give incorrect results for larger values if the class group has many cyclic factors. We thus recommend to double-check results using the function quadclassunit [...]

Thus, Sage should probably default to `quadclassunit()` and only use `qfbclassno()` if `proof=False` is set.

CC:  @tscholl2 @roed314

Component: **number fields**

_Issue created by migration from https://trac.sagemath.org/ticket/34801_

**Comment by b1acktothefuture** [comment]
Is this already taken care of here: #36184?

**Comment by yyyyx4** [comment]
No: The check there follows the *recommendation* in the [PARI documentation](https://pari.math.u-bordeaux.fr/dochtml/html/Arithmetic_functions.html) for when to use `quadclassunit()` instead of `qfbclassno()`, but for the latter there still is a warning that the results could be incorrect even in a nonempty part of the range where the documentation recommends it should be used. In contrast, for `quadclassunit()` the PARI documentation promises that the results are correct assuming GRH.

**Comment by Yashagarwal9798** [comment]
Hi @yyyyx4,
I’d be happy to work on this issue if it’s still open. Could you please assign it to me?

**Comment by yyyyx4** [comment]
Done, but note that in general you can just start working on things and submit a pull request when it's done. Given that this issue has been essentially idle for more than 3 years, it would seem very unlikely that anyone else randomly starts working on it at the same time. 🙂

**Comment by Yashagarwal9798** [comment]
Hi @yyyyx4,
I’ve opened a PR for this issue: https://github.com/sagemath/sage/pull/41320 .
I’d appreciate it if you could review it when you have a chance. Thanks!
